bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v3/product/ProductImpl.java

```diff
@@ -153,4 +178,87 @@ protected Variant mapVariant(ConfigurableVariant variant) {
         return visibleSectionsSet;
     }
+
+    public JSONArray fetchVariantsAsJsonArray() throws JSONException {
+        List<Variant> variants = getVariants();
+        JSONArray jsonArray = new JSONArray();
+
+        for (Variant variant : variants) {
+
+            LinkedHashMap<String, Object> variantMap = new LinkedHashMap<>();
+            LinkedHashMap<String, Object> variantMapWithNoSpecialPrice = new LinkedHashMap<>();
+            JSONArray assets = new JSONArray();
+
+            // Process assets
+            for (Asset asset : variant.getAssets()) {
+                JSONObject jsonAsset = new JSONObject();
+                jsonAsset.put("path", asset.getPath());
+                assets.put(jsonAsset);
+            }
+
+            Price priceRange = variant.getPriceRange();
+            // Set common fields for both cases
+            variantMap.put("@type", "Offer");
+            variantMap.put("sku", variant.getSku());
+            variantMap.put("url", getCanonicalUrl());
+            variantMap.put("image", assets.length() > 0 ? assets.getJSONObject(0).getString("path") : "");
+            variantMap.put("priceCurrency", priceRange != null ? priceRange.getCurrency() : "");
+
+            // Case when there is no special price
+            if (variant.getSpecialPrice() == null && variant.getSpecialToDate() == null) {
+                // For variants with no special price
+                variantMapWithNoSpecialPrice.putAll(variantMap);  // Copy common fields
+                variantMapWithNoSpecialPrice.put("price", priceRange.getRegularPrice());  // Set regular price
+                jsonArray.put(new JSONObject(variantMapWithNoSpecialPrice));  // Add to the array
+            } else {
+                // Case when there's a special price
+                variantMap.put("availability", variant.getInStock() ? "InStock" : "OutOfStock");
+
+                // Price specification
+                JSONObject priceSpecification = new JSONObject();
+                priceSpecification.put("@type", "UnitPriceSpecification");
+                priceSpecification.put("priceType", "https://schema.org/ListPrice");
+                if (priceRange != null) {
+                    priceSpecification.put("price", priceRange.getRegularPrice());
+                    priceSpecification.put("priceCurrency", priceRange.getCurrency());
+                }
+                variantMap.put("priceSpecification", priceSpecification);
+
+                // Set special price and date
+                variantMap.put("price", variant.getSpecialPrice());
+                variantMap.put("SpecialPricedate", variant.getSpecialToDate());
+
+                // Add the variant with special price to the array
+                jsonArray.put(new JSONObject(variantMap));
+            }
+        }
+
+        return jsonArray;
+    }
+
+    @Override
+    public String generateProductJsonLDString() throws JsonProcessingException, JSONException {
```

Comment:
You may consider instead of throwing these exceptions , to cache them, log them as error and return null.